api/metric/mock_test.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric
+
+import (
+	"context"
+
+	"go.opentelemetry.io/api/core"
+)
+
+type (
+	mockHandle struct {
+		descriptor *Descriptor
+		labelSet   *mockLabelSet
+	}
+
+	mockLabelSet struct {
+		meter  *mockMeter
+		labels map[core.Key]core.Value
+	}
+
+	batch struct {
+		ctx          context.Context
+		labelSet     *mockLabelSet
+		measurements []Measurement
+	}
+
+	observerData struct {
+		observer Observer
+		callback ObserverCallback
+	}
+
+	observerMap map[DescriptorID]observerData
+
+	mockMeter struct {
+		measurementBatches []batch
+		observers          observerMap
+	}
+)
```

Comment:
I wonder if it should be in a test-support package, not an internal one, so that users can write mock tests of their own easily. Either way, let's leave a TODO behind and not block this PR.

Reply:
opened #203.

Reply:
Actually, that was the first thing I did. But then I got import cycles (tests in api/metric package were importing internal/metric package to get the mock meter, and that in turn imported api/metric to implement the Meter interface). So annoying. So to resolve that I would either need to make tests a different package or put the mock meter into api/metric. I went with the latter for now.

Reply:
We can move the tersts in `api/metric` out of the "metric" package, into "metrics_test", to break this cycle. Let's leave it for a future TODO.